tests/all/epoch_interruption.rs

```diff
@@ -0,0 +1,124 @@
+        RawWaker::new(ptr, &VTABLE)
+    }
+
+    unsafe fn wake(ptr: *const ()) {
+        assert_eq!(ptr as usize, 5);
+    }
+
+    unsafe fn wake_by_ref(ptr: *const ()) {
+        assert_eq!(ptr as usize, 5);
+    }
+
+    unsafe fn drop(ptr: *const ()) {
+        assert_eq!(ptr as usize, 5);
+    }
+}
+
+fn build_engine() -> Engine {
+    let mut config = Config::new();
+    config.async_support(true);
+    config.epoch_interruption(true);
+    Engine::new(&config).unwrap()
+}
+
+fn make_env(engine: &Engine) -> Linker<()> {
+    let mut linker = Linker::new(engine);
+    let engine = engine.clone();
+
+    linker
+        .func_new(
+            "",
+            "bump_epoch",
+            FuncType::new(None, None),
+            move |_caller, _params, _results| {
+                engine.increment_epoch();
+                Ok(())
+            },
+        )
+        .unwrap();
+
+    linker
+}
+
+fn run_and_count_yields(wasm: &str, initial: u64, delta: u64) -> usize {
+    let engine = build_engine();
+    let linker = make_env(&engine);
+    let module = Module::new(&engine, wasm).unwrap();
+    let mut store = Store::new(&engine, ());
+    let instance = linker.instantiate(&mut store, &module).unwrap();
+    let f = instance.get_func(&mut store, "run").unwrap();
+
+    store.set_epoch_deadline_refill_delta(delta);
+    store.refill_epoch_deadline(initial);
+
+    let mut future = Box::pin(f.call_async(&mut store, &[], &mut []));
+    let mut yields = 0;
+    loop {
+        match future
+            .as_mut()
+            .poll(&mut Context::from_waker(&dummy_waker()))
+        {
+            Poll::Pending => {
+                yields += 1;
+            }
+            Poll::Ready(result) => {
+                result.unwrap();
+                break;
+            }
+        }
+    }
+
+    yields
+}
+
+#[test]
+fn epoch_yield_at_func_entry() {
+    assert_eq!(
+        1,
+        run_and_count_yields(
+            // Should yield at start of call to func 2.
+            "
+            (module
+                (import \"\" \"bump_epoch\" (func))
+                (func (export \"run\")
+                    call 0  ;; bump epoch
+                    call 2) ;; call func; will notice new epoch and yield
+                (func))
+            ",
+            1,
+            1,
+        )
+    );
+}
+
+#[test]
+fn epoch_yield_at_loop_header() {
+    assert_eq!(
+        2,
+        run_and_count_yields(
+            // Should yield at top of loop, once per five iters.
+            "
+            (module
+                (import \"\" \"bump_epoch\" (func))
+                (func (export \"run\")
+                    (local $i i32)
+                    (local.set $i (i32.const 10))
+                    (loop $l
+                        call 0
+                        (br_if $l (local.tee $i (i32.sub (local.get $i) (i32.const 1)))))))
+            ",
+            0,
+            5,
+        )
+    );
+}
```

Comment:
I think it'd be good to add a few more tests for epochs such as:

* Infinite loops can be interrupted from another thread
* Function entries can be interrupted from another thread (I think fuel has this test?)
* Something which exercises the double-check block, such as a function which calls another which changes the epoch but only one yield happens instead of multiple as it returns back up the stack or otherwise hits other functions.
* Dropping an in-progress future which is yielded on an epoch doesn't execute any further wasm.
* Initially a store's epoch deadline is zero and a yield immediately happens

Reply:
Added all of these -- thanks!